File: 0051-n-queens/0051-n-queens.java
class Solution {
    public List<List<String>> solveNQueens(int n) {
     boolean board[][]=new boolean[n][n];
    // nqueens b=new nqueens();
     List<List<String>> him=new ArrayList<>();
    queens(board, 0,him);
        return him;
    }
     void queens(boolean board[][],int row,    List<List<String>> him){
        if(row==board.length){
            display(board,him);
            return;
        }
        for(int col=0;col<board.length;col++){
            if(issafe(board,row,col)){
                board[row][col]=true;
                queens(board, row+1,him);
                board[row][col]=false;
            }
        }
    }
    private boolean issafe(boolean[][] board, int row, int col) {
      for(int i=0;i<row;i++){
        if(board[i][col]){
            return false;
        }
    }
        int maxleft=Math.min(row,col);
        for(int i=1;i<=maxleft;i++){
            if(board[row-i][col-i]){
                return false;
            }
        }
        int maxright=Math.min(row, board.length-col-1);
        for(int i=1;i<=maxright;i++){
            if(board[row-i][col+i]){
                return false;
            }
        }
        return true;
      }
    
    private void display(boolean[][] board, List<List<String>> him) {
        ArrayList<String> him1=new ArrayList<>();
        
        for (boolean[] bs : board) {
            String p="";
            for (boolean b: bs) {
                if(b){
                    p=p+'Q';
                }
                else{
                    p=p+'.';
                }
             
            }
                him1.add(p);
        }
       
        him.add(him1);
    }
}

File: 0051-n-queens/0051-n-queens.py
class Solution:
    def __init__(self):
        self.a=[]

    def solveNQueens(self, n: int) -> List[List[str]]:
        board=[[0]*n for i in range (n)]
        self.solve(board,0)
        return self.a
    def solve(self,b,r):
        if(r==len(b)):
            self.got(b)
            return
        for i in range(len(b)):
            if self.isSafe(b,r,i):
                b[r][i]=1
                self.solve(b,r+1)
                b[r][i]=0
    def got(self,b):
        l=[]
        
        for i in range(len(b)):
            st=[]
            for j in range(len(b)):
                st.append('Q' if b[i][j]==1 else ".")
            l.append("".join(st))
        self.a.append(l)
    def isSafe(self, b, i, j):
        n = len(b)

        # check row
        for c in range(n):
            if c == j:
                continue
            if b[i][c]:
                return False

        # check column
        for r in range(n):
            if r == i:
                continue
            if b[r][j]:
                return False

        # down-right diagonal
        r, c = i + 1, j + 1
        while r < n and c < n:
            if b[r][c]:
                return False
            r += 1
            c += 1

        # up-left diagonal
        r, c = i - 1, j - 1
        while r >= 0 and c >= 0:
            if b[r][c]:
                return False
            r -= 1
            c -= 1

        # down-left diagonal
        r, c = i + 1, j - 1
        while r < n and c >= 0:
            if b[r][c]:
                return False
            r += 1
            c -= 1

        # up-right diagonal
        r, c = i - 1, j + 1
        while r >= 0 and c < n:
            if b[r][c]:
                return False
            r -= 1
            c += 1

        return True


        

        